Added UploadStudentCsv component and tests (Closes #469)

Vraj67 opened the pull request:
This PR implements the UploadStudentCsv component and adds full test coverage. The component enables users to upload a .csv file, validates the file type, shows appropriate errors.

Features Implemented

- "Select Student CSV" button triggers hidden file input
- Only accepts .csv files.
- Shows file name in read-only text input
- Displays error message for invalid or missing files
- Upload button appears only when valid file is selected

 Unit Tests

- Renders all key inputs and buttons
- Validates proper file name rendering
- Handles non-CSV files with error
- Prevents upload without selected file
- Accepts .CSV uppercase extensions
- Allows re-selecting same file
- Handles empty file input correctly

Linked Issue

Closes #469

Vrajp08 committed:
Added UploadStudentCsv component and tests (Closes #469)

Vrajp08 committed:
Updated StudentCsv upload logic based on parse logic

achow-conestogac commented:
hi @Vraj67 i have many issues with this PR.

1. it looks like you only handle student data.  what will the people dealing with evaluations do? and what about any further parsing that is required?

achow-conestogac commented:
> Hi Professor @achow-conestogac Thank you for your feedback.
> 
> Regarding the concerns raised, the evaluation data is being handled by @Vrajpatel-new, and you can find the implementation in his pull request #515 . As you mentioned, the UploadEvaluationCsv component is directly linked with the evaluation data, and its purpose is specifically to support that functionality.

i don't think so.

firstly, your branch seems to fail when im trying to test it.  so this is something you need to address.

but let's get back to your flow.

this line here `const data = await parseCsv<Student>(selectedFile, "Student");` is where the data parsing is happening.

after it's called, you do `setStudents(data);`

which is a state here: `const [students, setStudents] = useState<Student[]>([]);`

so to summarize, your component looks like this

```const UploadStudentCsv: React.FC = () => {
  const [students, setStudents] = useState<Student[]>([]);
  ...

  const handleUpload = useCallback(async () => {
    setStudents([]);
    setErrorMessage(null);
    if (selectedFile) {
      try {
        const data = await parseCsv<Student>(selectedFile, "Student");
        setStudents(data);
   ...

  return (
  ..ui stuff
```

since you're setting the student data INSIDE this component, there's no way for another component to use the data that you have parsed. it's stuck as a state inside `UploadStudentCsv`.

This is why I have been skeptical of this piece of work since the beginning.

if i am wrong, please write me a test that tests this flow.

you will need to:
- implement a complete separate component that will require the parsed data.
- upload a csv with your component, then some how pass the data into the test component from bullet point 1.

Vrajp08 committed:
Implemented a complete separate component

Vraj67 commented:
Hi Professor @achow-conestogac 

Thank you for your feedback  you’re absolutely right.

Currently, the parsed data is trapped inside the UploadStudentCsv component as local state, which prevents other components from accessing it.

To fix this, I’ve refactored the flow using the prop-drilling pattern:

What I Changed:

- Removed internal state from UploadStudentCsv.

- Added a callback prop (onStudentsParsed) that sends the parsed data up to the parent.

Created a parent component CsvUploadDemo that:

- Holds the parsed students in state.

- Passes the data to another component (StudentSummary).

Created a completely separate component StudentSummary that:

Receives the parsed data as a prop.

Implement a complete separate component that will require the parsed data

## **How They Work Together**

- `CsvUploadDemo` manages the state and data flow.
- `UploadStudentCsv` parses the CSV and passes the data up.
- `StudentSummary` receives the data and displays it.

Displays the data for confirmation.

Vrajp08 committed:
Replaced Student-Specific Component with a Generic Solution

Vrajp08 committed:
Updated index file

Vraj67 commented:
Hi professor @achow-conestogac 
Thank you for your feedback. I've made the necessary changes to address all the concerns raised:

Changes Based on Your Feedback:
1. Replaced Student Specific Component with a Generic Solution

-  Removed: UploadStudentCsv.tsx and its corresponding test file

-  Introduced: CsvUploadButton a generic, reusable CSV upload component

- Renamed test file to CsvUploadButton.test.tsx

2. Moved Demo Components Into the Test File

- StudentSummary, CsvUploadDemo, and a new GenericDemo are now embedded in the test file

- No separate demo component files everything is consolidated and self-contained

3. Enhanced Generic Component Demonstration

-  Demonstrates support for multiple data types (e.g., Student, Evaluation)

 No code changes required to support new types.

Next Steps for Teams:
Replace any custom CSV upload components with CsvUploadButton

Follow the usage pattern shown in the test file for new or existing use cases

No need to build new upload components

achow-conestogac commented on src/ParseCsvService/csvUploader/CsvUploadButton.tsx:
```diff
@@ -0,0 +1,78 @@
+import { parseCsv } from "../ParseCsv";
+
+interface CsvUploadButtonProps<T> {
+  parseType: "Student" | "Evaluation";
```
If a person is required to give you a `parseType` then this is not generic.

At this point, I suggest abandoning this PR as I am not confident that this is a good idea.  This implementation assumes that the parser is only ever going to be used as a button and nothing else.
